### ISSUE TITLE
CSV dataset loading error

### ISSUE THREAD

**Issue by kboseong** [xuannianz/EfficientDet#10]
Thank you for your great work.

I constructed my own dataset with csv file like under

-----------------------------------------------

filename, xmin, ymin, xmax, ymax, class_name
...

-----------------------------------------------

training looks like working well, but there are too many warning messages like under 

UserWarning: Image with id 745 (shape (896, 896, 3)) contains no valid boxes after transform image.shape

b0 backbone is more often show the message, and b3 backbone has just few message
Can I know what it means?

**Comment by florisdesmedt** [comment]
walking through the code, this message is triggered if the annotations become too small (<10px in either width or height). Note that the network does not keeps the image at original resolution, but processes it at a resolution in relation to the `phi` value. The smaller the resolution, the larger the chance the annotations become too small. In B0 backbone this will happen more, since images (and annotations) are resized to 512px, compared to 896px in B3.
I suppose next to the resolution of the backbone, also certain augmentations could trigger this warning.

**Comment by 2696120622** [comment]
Can you trian the model with csv dataset successfully?

**Comment by kboseong** [comment]
> walking through the code, this message is triggered if the annotations become too small (<10px in either width or height). Note that the network does not keeps the image at original resolution, but processes it at a resolution in relation to the `phi` value. The smaller the resolution, the larger the chance the annotations become too small. In B0 backbone this will happen more, since images (and annotations) are resized to 512px, compared to 896px in B3.
> I suppose next to the resolution of the backbone, also certain augmentations could trigger this warning.

Thank you fore your reply:)

**Comment by kboseong** [comment]
> Can you trian the model with csv dataset successfully?

Yes, It works well. 
I think csv dataset is the fastest way for making your own dataset.
However, there is no code for inference with csv dataset, so you have to make your own script for inference your result.(valuation step is well supported )
If you have some questions about training with your own csv dataset, please reply here.

**Comment by 2696120622** [comment]
Can the model size be reduced by converting the training model to a inference model like in fizyr/keras-retinanet?

**Comment by balag59** [comment]
I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts  for the mAP to appear correctly during validation?

**Comment by 2696120622** [comment]
> I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?

I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation

**Comment by balag59** [comment]
Thanks, I'll give that a go.

**Comment by balag59** [comment]
> > I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?
> 
> I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation

Unfortunately, the issue still persists so I'm trying to figure out where I'm going wrong. Is there anything else that you modified for training a CSV dataset other than running it with a CSV instead of COCO/Pascal? I haven't modified anything else and even after 3 epochs the mAP is still 0 for all classes.

**Comment by 2696120622** [comment]
> > > I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?
> > 
> > 
> > I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation
> 
> Unfortunately, the issue still persists so I'm trying to figure out where I'm going wrong. Is there anything else that you modified for training a CSV dataset other than running it with a CSV instead of COCO/Pascal? I haven't modified anything else and even after 3 epochs the mAP is still 0 for all classes.

Sure, I haven't modified any evaluation functions/scripts. I have trained EfficientDet-D1 well with csv dataset following the step1 and step2. In step2, I reduce the learning rate to 1e-5. The validation mAP goes down each epoch.

**Comment by balag59** [comment]
> > > > I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?
> > > 
> > > 
> > > I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation
> > 
> > 
> > Unfortunately, the issue still persists so I'm trying to figure out where I'm going wrong. Is there anything else that you modified for training a CSV dataset other than running it with a CSV instead of COCO/Pascal? I haven't modified anything else and even after 3 epochs the mAP is still 0 for all classes.
> 
> Sure, I haven't modified any evaluation functions/scripts. I have trained EfficientDet-D1 well with csv dataset following the step1 and step2. In step2, I reduce the learning rate to 1e-5. The validation mAP goes down each epoch.

I'm sorry but I'm unable to resolve this still. Any chance that the input image size has to be 640x640 only? Or can I train with images of any size? I tried with 2 different csv datasets and I get 0 mAP for both in validation.

**Comment by 2696120622** [comment]
> > > > > I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?
> > > > 
> > > > 
> > > > I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation
> > > 
> > > 
> > > Unfortunately, the issue still persists so I'm trying to figure out where I'm going wrong. Is there anything else that you modified for training a CSV dataset other than running it with a CSV instead of COCO/Pascal? I haven't modified anything else and even after 3 epochs the mAP is still 0 for all classes.
> > 
> > 
> > Sure, I haven't modified any evaluation functions/scripts. I have trained EfficientDet-D1 well with csv dataset following the step1 and step2. In step2, I reduce the learning rate to 1e-5. The validation mAP goes down each epoch.
> 
> I'm sorry but I'm unable to resolve this still. Any chance that the input image size has to be 640x640 only? Or can I train with images of any size? I tried with 2 different csv datasets and I get 0 mAP for both in validation.

You can train with raw image of any size.

**Comment by balag59** [comment]
> > > > > > I'm able to train the model with a custom CSV dataset and can see the loss going down each epoch. However, the validation mAP just remains at 0 and I'm guessing that this has something to do with how it is evaluated for a custom dataset. Or am I missing something here? For those who trained with csv, did you have to modify any evaluation functions/scripts for the mAP to appear correctly during validation?
> > > > > 
> > > > > 
> > > > > I have trained with csv using phi 1. I have not modified any evaluation functions/scripts. The validation mAP appear correctly during validation
> > > > 
> > > > 
> > > > Unfortunately, the issue still persists so I'm trying to figure out where I'm going wrong. Is there anything else that you modified for training a CSV dataset other than running it with a CSV instead of COCO/Pascal? I haven't modified anything else and even after 3 epochs the mAP is still 0 for all classes.
> > > 
> > > 
> > > Sure, I haven't modified any evaluation functions/scripts. I have trained EfficientDet-D1 well with csv dataset following the step1 and step2. In step2, I reduce the learning rate to 1e-5. The validation mAP goes down each epoch.
> > 
> > 
> > I'm sorry but I'm unable to resolve this still. Any chance that the input image size has to be 640x640 only? Or can I train with images of any size? I tried with 2 different csv datasets and I get 0 mAP for both in validation.
> 
> You can train with raw image of any size.

Thanks! But I'm getting 0 mAP for 2 different csv datasets. Does this input format look right?
img_path,x1,y1,x2,y2,class_name

Could you also please tell me as to what snapshot you used to begin training? Was it imagenet? My model is not making any predictions at all and all classes are predicted as -1 for some reason.

**Comment by 2696120622** [comment]
I begin with imagenaet using --snapshot imagenet.

**Comment by balag59** [comment]
> I begin with imagenaet using --snapshot imagenet.

Funny, I'm using the same one. Yet I can't figure out as to why any dataset that I use the mAP is always 0. Especially considering that you have been able to get it working without modifying anything. Can you think of anything else that might be the problem here? The loss goes down no problem but the mAP is 0 for all classes after several epochs

**Comment by 2696120622** [comment]
> > I begin with imagenaet using --snapshot imagenet.
> 
> Funny, I'm using the same one. Yet I can't figure out as to why any dataset that I use the mAP is always 0. Especially considering that you have been able to get it working without modifying anything. Can you think of anything else that might be the problem here? The loss goes down no problem but the mAP is 0 for all classes after several epochs

Have you make csv annotations file or classes file correctly?

**Comment by balag59** [comment]
> > > I begin with imagenaet using --snapshot imagenet.
> > 
> > 
> > Funny, I'm using the same one. Yet I can't figure out as to why any dataset that I use the mAP is always 0. Especially considering that you have been able to get it working without modifying anything. Can you think of anything else that might be the problem here? The loss goes down no problem but the mAP is 0 for all classes after several epochs
> 
> Have you made csv annotations file or classes file correctly?

Here is the train annotation:
https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/train_annotations.csv

Here is the classes file:
https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/labels.csv

Do they seem ok?

**Comment by 2696120622** [comment]
> > > > I begin with imagenaet using --snapshot imagenet.
> > > 
> > > 
> > > Funny, I'm using the same one. Yet I can't figure out as to why any dataset that I use the mAP is always 0. Especially considering that you have been able to get it working without modifying anything. Can you think of anything else that might be the problem here? The loss goes down no problem but the mAP is 0 for all classes after several epochs
> > 
> > 
> > Have you made csv annotations file or classes file correctly?
> 
> Here is the train annotation:
> https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/train_annotations.csv
> 
> Here is the classes file:
> https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/labels.csv
> 
> Do they seem ok?

It looks like that there is nothing wrong in train_annotations.csv and labels.csv. What is your traing order?

**Comment by balag59** [comment]
> > > > > I begin with imagenaet using --snapshot imagenet.
> > > > 
> > > > 
> > > > Funny, I'm using the same one. Yet I can't figure out as to why any dataset that I use the mAP is always 0. Especially considering that you have been able to get it working without modifying anything. Can you think of anything else that might be the problem here? The loss goes down no problem but the mAP is 0 for all classes after several epochs
> > > 
> > > 
> > > Have you made csv annotations file or classes file correctly?
> > 
> > 
> > Here is the train annotation:
> > https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/train_annotations.csv
> > Here is the classes file:
> > https://github.com/priya-dwivedi/aerial_pedestrian_detection/blob/master/labels.csv
> > Do they seem ok?
> 
> It looks like that there is nothing wrong in train_annotations.csv and labels.csv. What is your training order?

I'm just doing step 1 of training as suggested:
python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --freeze-bn --batch-size 1 --steps 10 csv --val-annotations-path val_annotations.csv train_annotations.csv labels.csv

Am i missing anything?

**Comment by 2696120622** [comment]
I am sorry that I can't find anything incorrectly.
I train my csv using step 1 as following:
python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 68 csv csv/train_annotations.csv csv/train_classes.csv --val-annotations csv/val_annotations.csv.
I have not passed the --freeze-bn for a large batch-size. This goes well.

**Comment by balag59** [comment]
> I am sorry that I can't find anything incorrectly.
> I train my csv using step 1 as following:
> python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 68 csv csv/train_annotations.csv csv/train_classes.csv --val-annotations csv/val_annotations.csv.
> I have not passed the --freeze-bn for a large batch-size. This goes well.

That's fine thank you so much for all your help! By the way may I know the version of TF and keras you are using?

**Comment by 2696120622** [comment]
> > I am sorry that I can't find anything incorrectly.
> > I train my csv using step 1 as following:
> > python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 68 csv csv/train_annotations.csv csv/train_classes.csv --val-annotations csv/val_annotations.csv.
> > I have not passed the --freeze-bn for a large batch-size. This goes well.
> 
> That's fine thank you so much for all your help! By the way may I know the version of TF and keras you are using?

tensorflow               1.15.0
Keras                    2.2.5      
Keras-Applications       1.0.8      
Keras-Preprocessing      1.1.0      
keras-vis                0.4.1

**Comment by balag59** [comment]
> > > I am sorry that I can't find anything incorrectly.
> > > I train my csv using step 1 as following:
> > > python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 68 csv csv/train_annotations.csv csv/train_classes.csv --val-annotations csv/val_annotations.csv.
> > > I have not passed the --freeze-bn for a large batch-size. This goes well.
> > 
> > 
> > That's fine thank you so much for all your help! By the way may I know the version of TF and keras you are using?
> 
> tensorflow 1.15.0
> Keras 2.2.5
> Keras-Applications 1.0.8
> Keras-Preprocessing 1.1.0
> keras-vis 0.4.1

Thanks a ton! I'll try with the same versions and see if I can get it to work.

**Comment by balag59** [comment]
> > > I am sorry that I can't find anything incorrectly.
> > > I train my csv using step 1 as following:
> > > python train.py --snapshot imagenet --phi 1 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 68 csv csv/train_annotations.csv csv/train_classes.csv --val-annotations csv/val_annotations.csv.
> > > I have not passed the --freeze-bn for a large batch-size. This goes well.
> > 
> > 
> > That's fine thank you so much for all your help! By the way may I know the version of TF and keras you are using?
> 
> tensorflow 1.15.0
> Keras 2.2.5
> Keras-Applications 1.0.8
> Keras-Preprocessing 1.1.0
> keras-vis 0.4.1

With these versions installed I can finally see non 0 values for mAP so thank you so much for that!
